### PR TITLE
Transition closures with RootViewController only, instead of AnyTransitionPerformer

### PR DESCRIPTION
**Pull request by pauljohanneskraft**
Planned for 1.4

This PR changes all custom Transitions, although deprecation information has been added and it is backwards compatible. In addition to that, presented(from:) is now called from the rootViewController of the superCoordinator instead of the superCoordinator's anyTransitionPerformer.

### CODE REVIEW

File: XCoordinator/Classes/TransitionProtocol.swift [hunk 1]
```diff
@@ -39,10 +39,15 @@ public protocol TransitionProtocol {
...
     ///         The completion handler. Make sure to call this, whenever the transition is completed.
     ///
+    @available(*, deprecated, renamed: "perform(on:with:completion:)")
     func perform<C: Coordinator>(options: TransitionOptions,
                                  coordinator: C,
                                  completion: PresentationHandler?) where C.TransitionType == Self
 
+    func perform(on rootViewController: RootViewController,
+                 with options: TransitionOptions,
+                 completion: PresentationHandler?)
```
Comment (pauljohanneskraft) [inline XCoordinator/Classes/TransitionProtocol.swift]: Documentation missing.

File: XCoordinator/Classes/TransitionPerformer+TabBar.swift [hunk 2]
```diff
@@ -44,15 +44,15 @@ extension TransitionPerformer where TransitionType.RootViewController: UITabBarC
...
 
     func select(index: Int, with options: TransitionOptions, animation: Animation?, completion: PresentationHandler?) {
 
-        rootViewController.animationDelegate?.animation = animation
-        assert(animation == nil || rootViewController.animationDelegate != nil, """
+        animationDelegate?.animation = animation
+        assert(animation == nil || animationDelegate != nil, """
         Animations do not work, if your rootViewController's delegate is not a TabBarAnimationDelegate.
         This assertion might fail, if you did not call super.generateRootViewController to generate your rootViewController,
         or you set another delegate on your rootViewController. To set another delegate of your rootViewController, have a look
```
Comment (pauljohanneskraft) [inline XCoordinator/Classes/TransitionPerformer+TabBar.swift]: Adapt to not use rootViewController but only viewController or something similar.

File: XCoordinator/Classes/Transition.swift [hunk 3]
```diff
@@ -127,12 +142,44 @@ extension Transition {
...
+            let transitionPerformer = AnyTransitionPerformer(
+                ViewControllerTransitionPerformer(rootViewController)
+            )
+            perform(options, transitionPerformer, completion)
+        }
+    }
+}
+
+@available(*, deprecated)
+class ViewControllerTransitionPerformer<RootViewController: UIViewController>: TransitionPerformer {
```
Comment (pauljohanneskraft) [inline XCoordinator/Classes/Transition.swift]: Put into separate file

File: XCoordinator/Classes/Transition.swift [hunk 4]
```diff
@@ -108,15 +108,30 @@ public struct Transition<RootViewController: UIViewController>: TransitionProtoc
...
+    ///
+    public func perform(on rootViewController: RootViewController, with options: TransitionOptions, completion: PresentationHandler?) {
+        _perform(rootViewController, options, completion)
     }
 }
 
 extension Transition {
 
+    @available(*, deprecated, renamed: "PerformClosure")
+    public typealias Perform = (TransitionOptions, AnyTransitionPerformer<Transition>, PresentationHandler?) -> Void
```
Comment (pauljohanneskraft) [inline XCoordinator/Classes/Transition.swift]: Add documentation

File: XCoordinator/Classes/Transition.swift [hunk 5]
```diff
@@ -108,15 +108,30 @@ public struct Transition<RootViewController: UIViewController>: TransitionProtoc
...
+    @available(*, deprecated, renamed: "perform(on:with:completion:)")
     internal func perform(options: TransitionOptions,
                           performer: AnyTransitionPerformer<Transition>,
                           completion: PresentationHandler?) {
-        _perform(options, performer, completion)
+        _perform(performer.rootViewController, options, completion)
+    }
+
+    ///
+    /// The method to perform a certain transition using a coordinator.
```
Comment (pauljohanneskraft) [inline XCoordinator/Classes/Transition.swift]: Adapt documentation to not use coordinator

File: XCoordinator/Classes/Transition.swift [hunk 6]
```diff
@@ -78,15 +77,15 @@ public struct Transition<RootViewController: UIViewController>: TransitionProtoc
     ///     - presentables:
     ///         The presentables this transition is putting into the view hierarchy, if specifiable.
     ///         These presentables are used in the deep-linking feature.
-    ///     - animation:
+    ///     - animationInUse:
     ///         The transition animation this transition is using during the transition, i.e. the present animation
     ///         of a presenting transition or the dismissal animation of a dismissing transition.
     ///         Make sure to specify an animation here to use your transition with the
     ///         `registerInteractiveTransition` method in your coordinator.
```
Comment (pauljohanneskraft) [inline XCoordinator/Classes/Transition.swift]: Add information about perform-closure.

File: XCoordinator/Classes/TransitionPerformer+Navigation.swift [hunk 7]
```diff
@@ -19,11 +19,11 @@ extension TransitionPerformer where TransitionType.RootViewController: UINavigat
         if let animation = animation {
             viewController.transitioningDelegate = animation
         }
-        assert(animation == nil || rootViewController.animationDelegate != nil, """
-        Animations do not work, if your rootViewController's delegate is not a NavigationAnimationDelegate.
-        This assertion might fail, if you did not call super.generateRootViewController to generate your rootViewController,
-        or you set another delegate on your rootViewController. To set another delegate of your rootViewController, have a look
-        at `NavigationCoordinator.delegate`.
+        assert(animation == nil || animationDelegate != nil, """
```
Comment (nasirky) [inline XCoordinator/Classes/TransitionPerformer+Navigation.swift]: third person

File: XCoordinator/Classes/TransitionPerformer+Navigation.swift [hunk 8]
```diff
@@ -32,21 +32,21 @@ extension TransitionPerformer where TransitionType.RootViewController: UINavigat
...
         if let animation = animation {
-            rootViewController.topViewController?.transitioningDelegate = animation
+            topViewController?.transitioningDelegate = animation
         }
-        assert(animation == nil || rootViewController.animationDelegate != nil, """
-        Animations do not work, if your rootViewController's delegate is not a NavigationAnimationDelegate.
-        This assertion might fail, if you did not call super.generateRootViewController to generate your rootViewController,
-        or you set another delegate on your rootViewController. To set another delegate of your rootViewController, have a look
-        at `NavigationCoordinator.delegate`.
+        assert(animation == nil || animationDelegate != nil, """
```
Comment (nasirky) [inline XCoordinator/Classes/TransitionPerformer+Navigation.swift]: third person

File: XCoordinator/Classes/TransitionPerformer+Navigation.swift [hunk 9]
```diff
@@ -73,11 +73,11 @@ extension TransitionPerformer where TransitionType.RootViewController: UINavigat
             viewControllers.last?.transitioningDelegate = animation
         }
         resetChildrenAnimations(holding: [animation])
-        assert(animation == nil || rootViewController.animationDelegate != nil, """
-        Animations do not work, if your rootViewController's delegate is not a NavigationAnimationDelegate.
-        This assertion might fail, if you did not call super.generateRootViewController to generate your rootViewController,
-        or you set another delegate on your rootViewController. To set another delegate of your rootViewController, have a look
-        at `NavigationCoordinator.delegate`.
+        assert(animation == nil || animationDelegate != nil, """
+        Animations do not work, if the navigation controller's delegate is not a NavigationAnimationDelegate.
```
Comment (nasirky) [inline XCoordinator/Classes/TransitionPerformer+Navigation.swift]: third person

File: XCoordinator/Classes/TransitionPerformer+Navigation.swift [hunk 10]
```diff
@@ -100,14 +100,15 @@ extension TransitionPerformer where TransitionType.RootViewController: UINavigat
...
+            topViewController?.transitioningDelegate = animation
             viewController.transitioningDelegate = animation
         }
-        assert(animation == nil || rootViewController.animationDelegate != nil, """
-        Animations do not work, if your rootViewController's delegate is not a NavigationAnimationDelegate.
-        This assertion might fail, if you did not call super.generateRootViewController to generate your rootViewController,
-        or you set another delegate on your rootViewController. To set another delegate of your rootViewController, have a look
-        at `NavigationCoordinator.delegate`.
+
+        assert(animation == nil || animationDelegate != nil, """
```
Comment (nasirky) [inline XCoordinator/Classes/TransitionPerformer+Navigation.swift]: ?

File: XCoordinator/Classes/TransitionPerformer+TabBar.swift [hunk 11]
```diff
@@ -6,24 +6,24 @@
...
 
-        rootViewController.animationDelegate?.animation = animation
-        assert(animation == nil || rootViewController.animationDelegate != nil, """
-        Animations do not work, if your rootViewController's delegate is not a TabBarAnimationDelegate.
-        This assertion might fail, if you did not call super.generateRootViewController to generate your rootViewController,
-        or you set another delegate on your rootViewController. To set another delegate of your rootViewController, have a look
-        at `TabBarCoordinator.delegate`.
+        animationDelegate?.animation = animation
+        assert(animation == nil || animationDelegate != nil,"""
+        Animations do not work, if the tabbar controller's delegate is not a TabBarAnimationDelegate.
```
Comment (nasirky) [inline XCoordinator/Classes/TransitionPerformer+TabBar.swift]: ?

File: XCoordinator/Classes/TransitionPerformer+TabBar.swift [hunk 12]
```diff
@@ -33,36 +33,36 @@ extension TransitionPerformer where TransitionType.RootViewController: UITabBarC
...
                 completion: PresentationHandler?) {
 
-        rootViewController.animationDelegate?.animation = animation
-        assert(animation == nil || rootViewController.animationDelegate != nil, """
-        Animations do not work, if your rootViewController's delegate is not a TabBarAnimationDelegate.
-        This assertion might fail, if you did not call super.generateRootViewController to generate your rootViewController,
-        or you set another delegate on your rootViewController. To set another delegate of your rootViewController, have a look
-        at `TabBarCoordinator.delegate`.
+        animationDelegate?.animation = animation
+        assert(animation == nil || animationDelegate != nil, """
```
Comment (nasirky) [inline XCoordinator/Classes/TransitionPerformer+TabBar.swift]: ?

File: XCoordinator/Classes/TransitionPerformer+TabBar.swift [hunk 13]
```diff
@@ -33,36 +33,36 @@ extension TransitionPerformer where TransitionType.RootViewController: UITabBarC
...
+        Animations do not work, if the tabbar controller's delegate is not a TabBarAnimationDelegate.
```
Comment (nasirky) [inline XCoordinator/Classes/TransitionPerformer+TabBar.swift]: ?

File: XCoordinator/Classes/TransitionProtocol.swift [hunk 14]
```diff
@@ -39,10 +39,22 @@ public protocol TransitionProtocol {
...
+    @available(*, deprecated, renamed: "perform(on:with:completion:)")
     func perform<C: Coordinator>(options: TransitionOptions,
                                  coordinator: C,
                                  completion: PresentationHandler?) where C.TransitionType == Self
 
+    ///
+    /// Performs a transition on the given viewController.
+    ///
+    /// - Warning:
+    ///     Do not call this method directly. Instead use your coordinator's `performTransition` method or trigger
```
Comment (nasirky) [inline XCoordinator/Classes/TransitionProtocol.swift]: ?

**Pull request merged**